test/unit/base/test_subscription_controller_u.py

```diff
@@ -0,0 +1,475 @@
+    assert 'does not exist' in error_message
+    assert 'Current subscriptions:' in error_message
+
+
+# Tests for _attempt_unsubscribing_repeated method retry logic.
+#
+# These tests cover the complex retry loop logic that handles WebSocket
+# unsubscription attempts with confirmation waiting and failure handling.
+@pytest.mark.parametrize("wait_until_results,retries,expected_result,expected_send_calls,expected_wait_calls", [
+    ([True], 5, True, 1, 1),                    # Success first try
+    ([False, False, True], 3, True, 3, 3),     # Success after retries
+    ([False, False], 2, False, 2, 2),          # Failure after max retries
+])
+def test_attempt_unsubscribing_repeated_retry_logic_integration(subscription_controller, monkeypatch, wait_until_results, retries, expected_result, expected_send_calls, expected_wait_calls):
+    # Arrange
+    test_channel = 'test_channel'
+    test_payload = 'unsubscribe_payload'
+    subscription_controller._subscription_retries = retries
+
+    # Mock only external dependencies - test real _send_payload behavior
+    mock_ws_send = MagicMock(return_value=True)
+    monkeypatch.setattr(subscription_controller, 'send', mock_ws_send)
+
+    mock_wait_until = MagicMock(side_effect=wait_until_results)
+    monkeypatch.setattr('ibind.base.subscription_controller.wait_until', mock_wait_until)
+
+    # Act - Test real retry logic and error handling in _send_payload
+    result = subscription_controller._attempt_unsubscribing_repeated(test_channel, test_payload)
+
+    # Assert
+    assert result is expected_result
+    assert mock_ws_send.call_count == expected_send_calls
+    assert mock_wait_until.call_count == expected_wait_calls
+
+
+# Tests for recreate_subscriptions method
+#
+# These tests cover the subscription recreation logic that handles restoring
+# inactive subscriptions after connection issues or system restarts.
+@pytest.mark.parametrize("scenario,subscribe_success,expected_inactive_count", [
+    ('all_active', True, 0),      # No inactive subscriptions to recreate
+    ('all_inactive', True, 2),    # All inactive subscriptions should be recreated
+    ('mixed_active_inactive', True, 1),  # Only inactive should be recreated
+])
+def test_recreate_subscriptions_basic_functionality_integration(subscription_controller, monkeypatch, common_subscription_sets, scenario, subscribe_success, expected_inactive_count):
+    # Arrange
+    initial_subscriptions = common_subscription_sets[scenario]
+    subscription_controller._subscriptions = initial_subscriptions
+
+    # Mock only external dependencies - test real subscribe behavior
+    mock_ws_send = MagicMock(return_value=subscribe_success)
+    monkeypatch.setattr(subscription_controller, 'send', mock_ws_send)
+
+    # Mock subscription processor to create predictable payloads
+    mock_processor = subscription_controller._subscription_processor
+    mock_processor.make_subscribe_payload = MagicMock(return_value='test_payload')
```

Comment:
similarly, `_subscription_processor` is already a mock, from `mock_processor`. We can just `subscription_controller._subscription_processor.make_subscribe_payload .return_value = 'test_payload'`